src/Switch-Arduino_MCP4728/main.py

```diff
@@ -0,0 +1,269 @@
+    """
+
+    def __init__(self) -> None:
+        EmptyDevice.__init__(self)
+
+        self.shortname = "Arduino MCP4728"
+        self.variables = []
+        self.units = []
+
+        self.port_manager = True
+        self.port_types = ["COM"]
+        self.port_properties = {
+            "EOL": "\n",
+            "timeout": 15,
+            "baudrate": 115200,
+        }
+
+        self.multi_pins = False
+        self.multi_mcp = False
+        self.max_voltage = None
+        self.addresses = []
+
+        self.unit = {
+            "Voltage in V": "V",
+            "Output in %": "%",
+        }
+
+    def set_GUIparameter(self):
+        return {
+            "Channel": ["0", "1", "2", "3", "all"],
+            "I2C Address": "0",
+            "SweepMode": ["Voltage in V", "Output in %"],
+            "Voltage reference": ["External", "Internal 2.048 V", "Internal 4.096 V"],
+            "External voltage in V": 5.0,
+        }
+
+    def get_GUIparameter(self, parameter={}):
+        # Handle output pins
+        channel = parameter["Channel"]
+        if channel == "all":
+            self.multi_pins = True
+        else:
+            self.pin = int(channel)
+
+        # Handle I2C addresses
+        self.addresses = []
+        address_list = parameter["I2C Address"].split(",")
+        if len(address_list) == 1:
+            self.addresses.append(int(parameter["I2C Address"]))
+        else:
+            for addr in address_list:
+                self.addresses.append(int(addr))
+            self.multi_mcp = True
+
+        self.sweepmode = str(parameter["SweepMode"])
+
+        # Reference can either be internal (2.048 V or 4.096 V) or external (custom V)
+        self.reference_voltage = str(parameter["Voltage reference"])
+        if self.reference_voltage == "External":
+            self.external_voltage = float(parameter["External voltage in V"])
+
+        # Set variables and units
+        if channel == "all" and self.multi_mcp:
+            channel_num = len(self.addresses) * 4
+        elif channel == "all" and not self.multi_mcp:
+            channel_num = 4
+        else:
+            channel_num = 1
+
+        self.variables = []
+        self.units = []
+        for n in range(channel_num):
+            # If single channel, use the correct pin number
+            if channel_num == 1:
+                n = self.pin
+            self.variables.append(f"Channel{n}")
+            self.units.append(self.unit[self.sweepmode])
+
+    """ here, semantic standard functions start that are called by SweepMe! during a measurement """
+
+    def connect(self):
+        # Set Name/Number of COM Port as key
+        instance_key = f"mcp4728_{self.port.port.port}"
+
+        if instance_key not in self.device_communication:
+            # Wait for Arduino initialization
+            self.port.read()
+            self.device_communication[instance_key] = "Connected"
```

Comment:
Don't we need to unset the device communication flag somewhere when the measurement ends?

And on another note, as the comment talks about the initialization of the Arduino, this code could as well go into the initialize() function of the driver.

Finally a question of understanding: I can only see the code that it is waiting for the Arduino to finish initialization. But at which point is the arduino told to turn on or initialize?

Reply:
I've updated the instance_key parameter and added the disconnect function to pop the communication flag.

As far as I know, the Arduino initializes as soon as the connection is established and runs its setup() function.

Reply:
The reason why I think the code should be in 'connect' is that the semantic meaning of 'connect' is that a driver must be ready to communicate afterwards. As the Arduino is just ready for communication after we receiver the initial message, it should be done in 'connect'. This affects also the use with pysweepme, where one needs to use the connect method once to make sure the the Arduino is ready.
In future, get_driver of pyssweepme could automatically call 'connect' as this is always needed to be able to start a communication.
The semantic function "initialize" is more meant to be used to setup an instrument to the right configuration.

Reply:
Unsetting the flag in self.device_communication is not needed because there is not further decision at the end of a measurement whether the key is in or not. At the beginning of the next measurement, self.device_communication is anyway emptied, so I would suggest to remove the removal of the indentifier key at the end of a measurement from the code as it would otherwise suggest it has any meaning and could be rather confusing for people that take the driver as example.

Reply:
in SweepMe! maybe, but what if someone wants to use the driver in pysweepme? Then they have to restart their application every time they want to re-establish the connection